src/main/java/org/ohdsi/authenticator/filter/JWTAuthenticationFilter.java

```diff
@@ -35,28 +37,50 @@ public JWTAuthenticationFilter(Authenticator authenticator) {
     @Override
     public void doFilter(ServletRequest servletRequest, ServletResponse servletResponse, FilterChain filterChain) throws IOException, ServletException {
 
+        HttpServletRequest httpRequest = (HttpServletRequest) servletRequest;
         try {
-            HttpServletRequest httpRequest = (HttpServletRequest) servletRequest;
-            String authorize = httpRequest.getHeader(AuthorizationUtils.AUTHORIZE_HEADER);
-            if (StringUtils.isNotBlank(authorize) && authorize.startsWith(AuthorizationUtils.BEARER_AUTH)) {
-                String jwtToken = AuthorizationUtils.getToken(authorize);
+            String jwtToken = getToken(httpRequest);
+            if (StringUtils.isNotBlank(jwtToken)) {
                 String username = authenticator.resolveUsername(jwtToken);
                 if (Objects.nonNull(username) && SecurityContextHolder.getContext().getAuthentication() == null) {
-                    UserDetails userDetails = new User(username, "", getAuthorities(username));
+                    UserDetails userDetails = getUserDetails(username);
                     UsernamePasswordAuthenticationToken authentication = new UsernamePasswordAuthenticationToken(userDetails, jwtToken,
                             userDetails.getAuthorities());
+                    onSuccessAuthentication(httpRequest, userDetails, authentication);
                     SecurityContextHolder.getContext().setAuthentication(authentication);
                 }
             }
             filterChain.doFilter(servletRequest, servletResponse);
         } catch (AuthenticationException e) {
-            logger.error("Authentication error:", e);
+            String method = httpRequest.getMethod();
+            if (!HttpMethod.OPTIONS.matches(method)) {
+                if (logger.isDebugEnabled()) {
+                    logger.debug("Authentication failed", e);
+                } else {
+                    logger.error(String.format("Authentication failed: %s, requested: %s %s", e.getMessage(),
+                            method, httpRequest.getRequestURI()));
+                }
+            }
             HttpServletResponse response = (HttpServletResponse) servletResponse;
-            response.sendError(HttpServletResponse.SC_FORBIDDEN);
+            response.sendError(HttpServletResponse.SC_UNAUTHORIZED);
         }
     }
 
-    private Collection<? extends GrantedAuthority> getAuthorities(String username) {
+    protected String getToken(HttpServletRequest httpRequest) {
+
+        String authorize = httpRequest.getHeader(AuthorizationUtils.AUTHORIZE_HEADER);
+        return AuthorizationUtils.getToken(authorize);
+    }
+
+    protected void onSuccessAuthentication(HttpServletRequest request, UserDetails userDetails, AbstractAuthenticationToken authentication) {
```

Comment:
There are `AuthenticationSuccessHandler` in spring. If you really need to keep this it will be logical  also to add onFailureAuthentication method

Reply:
Indeed, but Authenticator library is agnostic to security library. In contrast, there's Apache Shiro used in Atlas.